File: scripts/test-release-version-sync.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'yaml'

ROOT = File.expand_path('..', __dir__)
class ContractError < StandardError; end

def validate(package, platforms, lockfile, config)
  pins = package.fetch('optionalDependencies')
  locked = lockfile.fetch('importers').fetch('ferromark').fetch('optionalDependencies')
  updaters = config.fetch('packages').fetch('.').fetch('extra-files')
  raise ContractError, 'native platform packages and optional dependencies differ' unless pins.keys.sort == platforms.keys.sort

  platforms.each do |name, platform|
    version = package.fetch('version')
    unless platform.fetch('version') == version && pins.fetch(name) == version
      raise ContractError, "#{name}: native package and dependency must match #{version}"
    end
    entry = locked.fetch(name)
    unless entry.fetch('specifier') == version
      raise ContractError, "#{name}: pnpm lockfile specifier must match #{version}"
    end
    unless entry.fetch('version') == "link:npm/#{name.delete_prefix('ferromark-')}"
      raise ContractError, "#{name}: lockfile must resolve to its local workspace package"
    end
    expected = {
      'type' => 'yaml',
      'path' => 'node/pnpm-lock.yaml',
      'jsonpath' => "$.importers.ferromark.optionalDependencies['#{name}'].specifier"
    }
    unless updaters.include?(expected)
      raise ContractError, "#{name}: release-please must update its pnpm lockfile specifier"
    end
  end
end

def assert_rejected(label, inputs)
  copy = Marshal.load(Marshal.dump(inputs))
  yield(*copy)
  validate(*copy)
  abort "#{label}: invalid release configuration was accepted"
rescue ContractError
  # Expected: detect version drift without running any installs or builds.
end

package = JSON.parse(File.read(File.join(ROOT, 'node/ferromark/package.json')))
platforms = Dir.glob(File.join(ROOT, 'node/ferromark/npm/*/package.json')).to_h do |path|
  platform = JSON.parse(File.read(path))
  [platform.fetch('name'), platform]
end
lockfile = YAML.safe_load(File.read(File.join(ROOT, 'node/pnpm-lock.yaml')), aliases: false)
config = JSON.parse(File.read(File.join(ROOT, 'release-please-config.json')))
inputs = [package, platforms, lockfile, config]

begin
  validate(*inputs)
  if ARGV == ['--self-test']
    name = platforms.keys.first
    assert_rejected('outdated lockfile', inputs) do |_, _, lock, _|
      lock['importers']['ferromark']['optionalDependencies'][name]['specifier'] = '0.0.0'
    end
    assert_rejected('mismatched native package', inputs) do |_, native, _, _|
      native[name]['version'] = '0.0.0'
    end
    assert_rejected('mismatched dependency pin', inputs) do |pkg, _, _, _|
      pkg['optionalDependencies'][name] = '0.0.0'
    end
    assert_rejected('registry resolution', inputs) do |_, _, lock, _|
      lock['importers']['ferromark']['optionalDependencies'][name]['version'] = '0.0.0'
    end
    assert_rejected('missing release updater', inputs) do |_, _, _, cfg|
      cfg['packages']['.']['extra-files'].reject! { |entry| entry.is_a?(Hash) && entry['type'] == 'yaml' }
    end
  elsif !ARGV.empty?
    abort 'usage: test-release-version-sync.rb [--self-test]'
  end
rescue ContractError => error
  abort "Release version sync: #{error.message}"
end

puts "release version sync checks passed (#{platforms.length} native packages)"

File: scripts/test-release-version-sync.mjs
import assert from "node:assert/strict";
import { readdirSync } from "node:fs";
import { describe, it } from "node:test";

import {
  ContractError,
  deepCopy,
  parseYaml,
  readRepositoryFile,
  repositoryPath,
} from "./lib/contracts.mjs";

function failContract(message) {
  throw new ContractError(message);
}

function versionUpdater(updaters, path) {
  return updaters.some(
    (updater) =>
      typeof updater === "object" &&
      updater !== null &&
      updater.type === "json" &&
      updater.jsonpath === "$.version" &&
      (updater.path === path ||
        (updater.glob === true &&
          path.startsWith(updater.path.split("*")[0]) &&
          path.endsWith(updater.path.split("*")[1]))),
  );
}

function validate({ package: packageJson, platforms, lockfile, config }) {
  const references = packageJson.optionalDependencies;
  const locked = lockfile.importers.ferromark.optionalDependencies;
  const updaters = config.packages["."]["extra-files"];
  if (Object.keys(references).sort().join() !== Object.keys(platforms).sort().join()) {
    failContract("native platform packages and optional dependencies differ");
  }
  if (!versionUpdater(updaters, "node/ferromark/package.json")) {
    failContract("release-please must update the npm facade version");
  }
  if (updaters.some((updater) => updater?.path?.endsWith("pnpm-lock.yaml"))) {
    failContract("the pnpm lockfile is generated state, not a release template target");
  }

  for (const [name, platform] of Object.entries(platforms)) {
    const version = packageJson.version;
    if (platform.version !== version) {
      failContract(`${name}: native package must match ${version}`);
    }
    // The facade carries no sidecar version at all: `workspace:*` resolves to
    // the sidecar's own version while packing, so a release bumps nine
    // `version` fields and nothing else — the lockfile included.
    if (references[name] !== "workspace:*") {
      failContract(`${name}: must be referenced with the workspace protocol`);
    }
    const entry = locked[name];
    if (entry?.specifier !== "workspace:*") {
      failContract(`${name}: pnpm lockfile specifier must be the workspace protocol`);
    }
    if (entry.version !== `link:npm/${name.replace(/^ferromark-/, "")}`) {
      failContract(`${name}: lockfile must resolve to its local workspace package`);
    }
    if (
      !versionUpdater(
        updaters,
        `node/ferromark/npm/${name.replace(/^ferromark-/, "")}/package.json`,
      )
    ) {
      failContract(`${name}: release-please must update its native package version`);
    }
  }
}

function readInputs() {
  const packageJson = JSON.parse(readRepositoryFile("node", "ferromark", "package.json"));
  const platforms = {};
  for (const entry of readdirSync(repositoryPath("node", "ferromark", "npm")).sort()) {
    const platform = JSON.parse(
      readRepositoryFile("node", "ferromark", "npm", entry, "package.json"),
    );
    platforms[platform.name] = platform;
  }
  return {
    package: packageJson,
    platforms,
    lockfile: parseYaml(readRepositoryFile("node", "pnpm-lock.yaml")),
    config: JSON.parse(readRepositoryFile("release-please-config.json")),
  };
}

function assertRejected(mutate) {
  const inputs = deepCopy(readInputs());
  mutate(inputs);
  assert.throws(() => validate(inputs), ContractError);
}

const inputs = readInputs();
const platformName = Object.keys(inputs.platforms).sort()[0];

describe("release version sync", () => {
  it("keeps every published version in step", () => {
    validate(inputs);
    assert.equal(Object.keys(inputs.platforms).length, 8);
  });

  it("rejects a lockfile specifier that pins a version", () => {
    assertRejected((copy) => {
      copy.lockfile.importers.ferromark.optionalDependencies[platformName].specifier = "0.0.0";
    });
  });

  it("rejects a mismatched native package version", () => {
    assertRejected((copy) => {
      copy.platforms[platformName].version = "0.0.0";
    });
  });

  it("rejects a sidecar reference that pins a version", () => {
    assertRejected((copy) => {
      copy.package.optionalDependencies[platformName] = "0.0.0";
    });
  });

  it("rejects a platform package resolved from the registry", () => {
    assertRejected((copy) => {
      copy.lockfile.importers.ferromark.optionalDependencies[platformName].version = "0.0.0";
    });
  });

  it("rejects a missing release updater", () => {
    assertRejected((copy) => {
      copy.config.packages["."]["extra-files"] = copy.config.packages["."]["extra-files"].filter(
        (updater) => updater?.glob !== true,
      );
    });
  });

  it("rejects a release updater that writes into the pnpm lockfile", () => {
    assertRejected((copy) => {
      copy.config.packages["."]["extra-files"].push({
        type: "yaml",
        path: "node/pnpm-lock.yaml",
        jsonpath: "$.importers.ferromark.optionalDependencies[*].specifier",
      });
    });
  });
});
